Performance: Optimized data loading of course

Sh-AhsanArif opened the pull request:
- Optimized data loading on app dashboard screen (cached data first)
- Optimized data loading on course outline screen (cached data first)
- Minor code refactoring

Commit:
perf: optimised data loading on dashboard and course details

openedx-webhooks commented:
Thanks for the pull request, @Sh-AhsanArif!


This repository is currently maintained by @openedx/openedx-mobile-maintainers.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary><h4>:radio_button: Get product approval</h4></summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details><details>
<summary><h4>:radio_button: Provide context</h4></summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details><details>

<summary><h4>:radio_button: Submit a signed contributor agreement (CLA)</h4></summary>

:warning: We ask all contributors to the Open edX project to submit a [signed contributor agreement](https://openedx.org/cla) or indicate their institutional affiliation.
Please see the [CONTRIBUTING](https://github.com/openedx/.github/blob/master/CONTRIBUTING.md) file for more information.

If you've signed an agreement in the past, you may need to re-sign.
See [The New Home of the Open edX Codebase](https://open.edx.org/blog/the-new-home-of-the-open-edx-codebase/) for details.

Once you've signed the CLA, please allow 1 business day for it to be processed.
After this time, you can re-run the CLA check by adding a comment below that you have signed it.
If the CLA check continues to fail, you can tag the `@openedx/cla-problems` team in a comment for further assistance.
</details><details>
<!-- comment:no_cla -->




<summary><h4>:radio_button: Get a green build</h4></summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<summary><h3>Where can I find more information? </h3></summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details><details>
<summary><h3>When can I expect my changes to be merged?</h3></summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

e0d commented:
@Sh-AhsanArif as you are covered by the Arbisoft entity CLA, I've added you to the triage team.  This will allow tests to run automatically on your PRs.  Please accept the invite.

Commit:
fix: course unit tests

PavloNetrebchuk commented:
@Sh-AhsanArif 
Hi, great job. If you are moving the logic to Flow, it would be great to make methods that do not return Flow private. What do you think?

PavloNetrebchuk commented:
@Sh-AhsanArif 
Okay, please resolve the conflicts and I'll approve

PavloNetrebchuk commented:
Hi, @Sh-AhsanArif 

I’ve reviewed the app again and have a few questions:
	1.	When opening a course in a fresh install of the app, we encounter a `NoCachedDataException`. Could you please take a look and fix this?
	2.	The `Result` layer seems unnecessary and inconsistent with the rest of the app. Since we typically use catch, it would be great if we could stick to this approach for consistency.
	3.	I have some doubts regarding the following code in `CourseContainerViewModel` and similar code in `CourseOutlineViewModel`:
```
courseFlow.take(1).combine(enrollmentFlow.take(1)) { course, enrollment ->
    course to enrollment
}	
```
Could you clarify why we are using `take(1)` here? Would it be possible to achieve the same result with:
```
courseFlow.combine(enrollmentFlow) { course, enrollment ->
    ...
}
```

omerhabib26 commented:
Hey @Sh-AhsanArif, thank you so much for this improvement! I had a chance to take a quick look, and I noticed a couple of small UI issues in your repo, as shown in the video:

There’s a glitch on the home screen when the user lands for the first time. IMO, it should show a loader instead.
The API seems to be crashing on the Course dashboard with the error org.openedx.core.exception.NoCachedDataException.

[OpenEdxPR.webm](https://github.com/user-attachments/assets/aa01853e-e546-4df2-8d74-352ad50ddfdb)

Commit:
fix: course load exception and removed ui result class

Commit:
fix: lint checks

Sh-AhsanArif commented:
Hey @PavloNetrebchuk, @omerhabib26 

Thank you for reviewing the PR and providing valuable feedback. I’ve incorporated the suggested changes and fixed the mentioned issues. Please have a look at the latest updates and let me know if any further adjustments are needed.
I appreciate your time and guidance!

PavloNetrebchuk commented on course/src/main/java/org/openedx/course/presentation/container/CourseContainerViewModel.kt:
```diff
@@ -171,46 +171,28 @@ class CourseContainerViewModel(
+                        else -> _courseAccessStatus.value = CourseAccessError.UNKNOWN
                     }
                 }
             } catch (e: Exception) {
```
And here

 ```
   fun fetchCourseDetails() {
        courseDashboardViewed()

        // If data is already loaded, do nothing
        if (_dataReady.value != null) return

        _showProgress.value = true

        viewModelScope.launch {
            val courseStructureFlow = interactor.getCourseStructureFlow(courseId)
                .catch { e ->
                    handleFetchError(e)
                    emit(null)
                }
            val courseDetailsFlow = interactor.getEnrollmentDetailsFlow(courseId)
                .catch { emit(null) }
            courseStructureFlow.combine(courseDetailsFlow) { courseStructure, courseEnrollmentDetails ->
                courseStructure to courseEnrollmentDetails
            }.catch { e ->
                handleFetchError(e)
            }.collect { (courseStructure, courseEnrollmentDetails) ->
                when {
                    courseEnrollmentDetails != null -> {
                        handleCourseEnrollment(courseEnrollmentDetails)
                    }

                    courseStructure != null -> {
                        handleCourseStructureOnly(courseStructure)
                    }

                    else -> {
                        _courseAccessStatus.value = CourseAccessError.UNKNOWN
                    }
                }
            }
        }
    }
```

Sh-AhsanArif commented on course/src/main/java/org/openedx/course/presentation/outline/CourseOutlineViewModel.kt:
```diff
@@ -184,47 +184,31 @@ class CourseOutlineViewModel(
+
+                    initializeCourseData(blocks, courseStructure, courseStatus, datesBannerInfo)
+                }
             } catch (e: Exception) {
```
Sounds good @PavloNetrebchuk. I planned it next but let me update it here now, will have to update the test cases as well accordingly.

Commit:
chore: moved to flow.catch from try/catch in course outline and container viewmodels

PavloNetrebchuk reviewed (approved):
Approve

omerhabib26 commented:
Thank you @Sh-AhsanArif for the update. I'll start its review on monday.

omerhabib26 commented on app/schemas/org.openedx.app.room.AppDatabase/1.json:
```diff
@@ -0,0 +1,772 @@
+      "INSERT OR REPLACE INTO room_master_table (id,identity_hash) VALUES(42, 'bcac519e74e751a75f3e6fa5d39ac5a3')"
+    ]
+  }
+}
```
Please add a line at EOF.

Sh-AhsanArif replied:
Hi @omerhabib26.
Thank you for reviewing the PR. The Schema files are auto-generated by Room, so it's best not to edit them manually to ensure schema consistency and avoid auto migration issues. Let me know your thoughts on this!

omerhabib26 commented on app/schemas/org.openedx.app.room.AppDatabase/2.json:
```diff
@@ -0,0 +1,978 @@
+      "INSERT OR REPLACE INTO room_master_table (id,identity_hash) VALUES(42, 'ed545aec6739ec7692c4bb72179331c4')"
+    ]
+  }
+}
```
same here

omerhabib26 reviewed (approved):
only a minor nit

miankhalid commented:
@volodymyr-chekyrta are we good to merge this PR? Pavlo & Omer both have approved this PR.

volodymyr-chekyrta commented:
> @volodymyr-chekyrta are we good to merge this PR? Pavlo & Omer both have approved this PR.

merging 👍
